### DISCUSSION TITLE
How do one interrogate a libp2p node for the protocols it supports

### DISCUSSION THREAD

**Discussion by finlaydotb** [libp2p/rust-libp2p#2674] [Q&A]
As said in the title, how do one interrogate a libp2p node for the protocols it supports?

Lets say I have a PeerId or a Multadr I want to communicate with, what requests do I need to make to negotiate the protocol it supports?

**Comment by finlaydotb** [comment]
Reading the documentation here: https://docs.libp2p.io/concepts/protocols/

> When dialing out to initiate a new stream, libp2p will send the protocol id of the protocol you want to use. The listening peer on the other end will check the incoming protocol id against the registered protocol handlers.
>
> If the listening peer does not support the requested protocol, it will end the stream, and the dialing peer can try again with a different protocol, or possibly a fallback version of the initially requested protocol.
>
> If the protocol is supported, the listening peer will echo back the protocol id as a signal that future data sent over the stream will use the agreed protocol semantics.

It seems there is no way to ask a node for the protocol it supports, and protocol negotiation is basically trial and error: dial a node with a protocol, if it ends the connection, then that protocol is not supported.

Is this a clear representation of what obtains? Or there are other ways to know the protocol a node supports?

**Comment by rkuhn** [comment]
The [identify protocol](https://docs.libp2p.io/concepts/protocols/#identify) transmits a list of protocol names supported by the peer (in addition to the purposes documented at the link), so it is a great idea to run that protocol on all libp2p nodes. The default settings include an initial delay for sending the identify info, but you can set that to zero to make the handshake quicker.
